src/axom/quest/MarchingCubesAlgo.cpp

```diff
@@ -0,0 +1,614 @@
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/quest/MarchingCubesAlgo.hpp"
+#include "axom/quest/detail/marching_cubes_lookup.hpp"
+#include "axom/mint/mesh/UnstructuredMesh.hpp"
+#include "conduit_blueprint.hpp"
+#include "axom/fmt.hpp"
+
+#ifndef __WHERE
+  #define __STRINGIZE(x) __STRINGIZE2(x)
+  #define __STRINGIZE2(x) #x
+  //!@brief String literal for code location
+  #define __WHERE \
+    __FILE__ ":" __STRINGIZE(__LINE__) "(" + std::string(__func__) + ") "
+#endif
+
+namespace axom
+{
+namespace quest
+{
+MarchingCubesAlgo::MarchingCubesAlgo(const conduit::Node& bpMesh,
+                                     const std::string& coordsetName,
+                                     const std::string& maskField)
+  : _sd()
+  , _ndim(0)
+  , _coordsetPath("coordsets/" + coordsetName)
+  , _fcnPath()
+  , _maskPath(maskField.empty() ? std::string() : "fields/" + maskField)
+  , _surfaceMesh(nullptr)
+  , _cellIdField()
+  , _domainIdField()
+{
+  _sd.reserve(conduit::blueprint::mesh::number_of_domains(bpMesh));
+  for(auto& dom : bpMesh.children())
+  {
+    _sd.emplace_back(new MarchingCubesAlgo1(dom, coordsetName, maskField));
+    if(_ndim == 0)
+    {
+      _ndim = _sd.back()->dimension();
+    }
+    else
+    {
+      SLIC_ASSERT(_ndim == _sd.back()->dimension());
+    }
+  }
+}
+
+void MarchingCubesAlgo::set_function_field(const std::string& fcnField)
+{
+  _fcnPath = "fields/" + fcnField;
+  for(auto& s : _sd)
+  {
+    s->set_function_field(fcnField);
+  }
+}
+
+/*!
+  @brief Set the output surface mesh object.
+*/
+void MarchingCubesAlgo::set_output_mesh(axom::mint::Mesh* surfaceMesh)
+{
+  _surfaceMesh = surfaceMesh;
+
+  for(auto& s : _sd)
+  {
+    s->set_output_mesh(_surfaceMesh);
+  }
+}
+
+void MarchingCubesAlgo::compute_iso_surface(double contourVal)
+{
+  SLIC_ASSERT_MSG(_surfaceMesh,
+                  "You must call set_output_mesh before compute_iso_surface.");
+  SLIC_ASSERT_MSG(
+    !_fcnPath.empty(),
+    "You must call set_function_field before compute_iso_surface.");
+
+  if(!_domainIdField.empty() &&
+     !_surfaceMesh->hasField(_domainIdField, axom::mint::CELL_CENTERED))
+  {
+    _surfaceMesh->createField<int>(_domainIdField, axom::mint::CELL_CENTERED);
+  }
+
+  for(int dId = 0; dId < _sd.size(); ++dId)
+  {
+    std::shared_ptr<MarchingCubesAlgo1>& single = _sd[dId];
+
+    auto nPrev = _surfaceMesh->getNumberOfCells();
+    single->compute_iso_surface(contourVal);
+    auto nNew = _surfaceMesh->getNumberOfCells();
+
+    if(nNew > nPrev && !_domainIdField.empty())
+    {
+      auto* domainIdPtr =
+        _surfaceMesh->getFieldPtr<int>(_domainIdField, axom::mint::CELL_CENTERED);
+      for(int n = nPrev; n < nNew; ++n)
+      {
+        domainIdPtr[n] = dId;
+      }
+    }
+  }
+}
+
+MarchingCubesAlgo1::MarchingCubesAlgo1(const conduit::Node& dom,
```

Comment:
Some basic class documentation describing what it does, how it works, etc. should be added. Same for member methods.

What's the difference between Algo1 and Algo above?

Reply:
See above.  I'll rename the classes and document as Kenny suggested.